data/gnome-screensaver-command.1
.\" Copyright (C) 2007 Sven Arvidsson <sa@whiz.se>
.\"
.\" This is free software; you may redistribute it and/or modify
.\" it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2,
.\" or (at your option) any later version.
.\"
.\" This is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\"You should have received a copy of the GNU General Public License along
.\"with this program; if not, write to the Free Software Foundation, Inc.,
.\"51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.TH gnome-screensaver-command 1 "2007\-09\-27" "GNOME"
.SH NAME
gnome-screensaver-command \- controls GNOME screensaver
.SH SYNOPSIS
.B gnome-screensaver-command
.RI [ OPTION... ]
.SH DESCRIPTION
gnome-screensaver-command is a tool for controlling an already
running instance of gnome-screensaver.
.SH OPTIONS
.TP
.B \-\-exit
Causes the screensaver to exit gracefully
.TP
.B \-q, \-\-query
Query the state of the screensaver
.TP
.B \-t, \-\-time
Query the length of time the screensaver has been active
.TP
.B \-l, \-\-lock
Tells the running screensaver process to lock the screen immediately
.TP
.B \-a, \-\-activate
Turn the screensaver on (blank the screen)
.TP
.B \-d, \-\-deactivate
If the screensaver is active then deactivate it (un-blank the screen)
.TP
.B \-V, \-\-version
Version of this application
.SH AUTHORS
.B gnome-screensaver-command
is written by William Jon McCann <mccann@jhu.edu>.
.P
This manual page was written by Sven Arvidsson <sa@whiz.se>.
.SH SEE ALSO
.BR "gnome-screensaver" (1)



NEWS
================
Version  2.91.91
================

- Follow the users clock format preference
- Port to GDBus
- Build cleanups
- Remove --poke option from the commandline tool (didn't work anyway)
- Remove unused delay and timeout preferences
- Remove inhibit interface (inhibit the session instead)
- Remove theme and job support
- Update style to match gnome-shell
- Use accountsservice to get face
- Translation updates (Korean, Simplified Chinese,
  Norwegian bokmål, Latvian Oriya, Ukrainian, Russian)

===============
Version  2.91.90
===============

- Use username in top panel if real name is empty
- Various settings fixes
- Translation updates

===============
Version  2.91.4
===============

- Fix build against recent GTK+
- GNOME 3 style updates
- Translation updates

===============
Version  2.91.3
===============

- Fix build against recent GTK+
- Use GIO to launch gdmflexiserver
- Translation updates

===============
Version  2.91.2
===============

- Fix build against recent GTK+
- Port to GSettings

===============
Version  2.91.1
===============

- Fix build against recent GTK+ and gnome-desktop
- Port background handling to gsettings
- Translation updates

===============
Version  2.91.0
===============

 - Switch to GTK+ 3.0
 - Fix crash on systems that lack vidmode and randr extensions
 - Add session service activation file
 - Use better unicode symbols in strings
 - Drop "Leave a message" feature
 - Drop unlock dialog theme support
 - Drop support for non-PAM based authentication
 - Improve aesthetics of unlock dialog
 - Monitor hotplug fixes
 - Auto-activate screensaver when needed if not running
 - Translation updates

===============
Version  2.30.0
===============

Translation updates

===============
Version 2.29.92
===============

Fixed bugs:
 - Build fix for systems with x11 < 1.3

===============
Version 2.29.91
===============

Fixed bugs:
 - #609789 - CVE-2010-0422, handle monitor hotplugging more securely
 - Fix screen dirt in floaters screensaver

===============
Version 2.29.90
===============

Fixed bugs:
 - #609337, CVE-2010-0414, handle monitor removal more securely

===============
Version 2.29.1
===============

Fixed bugs:
 - #597874, Use xrandr 1.2 gamma fading
 - RH bug #546656, Fix crash in popsquares
 - #598476, Don't crash when lock dialog dies during shake
 - #593616, Fix multi-head issues
 - #600488, Remove session inhibitors if the originator falls of the bus


===============
Version 2.28.0
===============

Fixed bugs:

 - #444927, get idle time from xorg
 - #463010, Port gnome-screensaver from libglade to gtkbuilder
 - #564061, Install .desktop file
 - #570402, screensaver crash if number of screens changes
 - #570941, String not marked for translation gnome-screensaver-command.c
 - #574896, [Leave message] Cancel button should clean clipboard buffe
 - #576463, can't unlock screen after upgrade until restarting gnome-
            screensaver
 - #581995, Fade out doesn't properly fall back to gamma number if X server
            only supports 0-size ramp
 - #586174, Encoding key in desktop files is deprecated
 - #589728, gnome-screensaver: implicit declaration g_printf
 - #590776, screensaver is never activated
 - #591193, Update/Create .gitignore
 - #591194, Add Comment and Icon in gnome-screensaver.desktop
 - #594082, gnome-screensaver ignores input when fading to black

Translators:

- Khaled Hosny (ar)
- Amitakhya Phukan (as)
- Alexander Shopov (bg)
- Runa Bhattacharjee (bn_IN)
- Jamil Ahmed (bn)
- Denis ARNAUD (br)
- Carles Ferrando (ca@valencia)
- Petr Kovar (cz)
- Ask H. Larsen (da)
- Mario Blättermann, Christian Kirbach (de)
- Philip Withnall (en_GB)
- Jorge González (es)
- Ivar Smolin, Mattias Põldaru (et)
- Inaki Larranaga Murgoitio (eu)
- Tommi Vainikainen (fi)
- Claude Paroz (fr)
- Seán de Búrca (ga)
- Antón Méixome (gl)
- Sweta Kothari (gu)
- Yaron Shahrabani (he)
- Rajesh Ranjan (hi)
- Gabor Kelemen (hu)
- Luca Ferretti (it)
- Shankar Prasad (kn)
- Changwoo Ryu (ko)
- Gintautas Miliauskas (lt)
- Rajesh Ranjan (mai)
- Ani (ml)
- Sandeep Shedmake (mr)
- Kjartan Maraas (nb)
- Niels-Christoph Fiedler (nds)
- Manoj Kumar Giri (or)
- A S Alam (pa)
- Tomasz Dominikowski (pl)
- Rodrigo L. M. Flores, Krix Apolinário (pt_BR)
- Duarte Loreto (pt)
- Lucian Adrian Grijincu (ro)
- Matej Urbančič (sl)
- Miloš Popović (sr@latin)
- Miloš Popović (sr)
- Daniel Nylander (sv)
- ifelix (ta)
- krishnababu k (te)
- Theppitak Karoonboonyanan (th)
- Maxim V. Dziumanenko (uk)
- Chao-Hsiung Liao (zh_HK)
- Chao-Hsiung Liao (zh_TW)

===============
Version 2.27.0
===============

- Add an autostart file for gnome-screensaver (#585485)

Updated Translations: Telugu, Catalan, zh_CN, Crimean Tatar

===============
Version 2.26.1
===============

- Fixed #572955, audit GError usage (William Jon McCann)
- Fixed #572956, use g_strerror (William Jon McCann)
- Fixed #573495, logout_command setting may be overwritten with keyboard_command value (William Jon McCann)
- Fixed #577739, gnome-screensaver-gl-helper calls gdk_x11_visual_get_xvisual with NULL visual (William Jon McCann)
- Fixed #578669, Screenserver (William Jon McCann)

Translations:
- Updated crh: Reşat SABIQ
- Updated ja: Takeshi AIHANA
- Updated kn: Shankar Prasad

===============
Version 2.26.0
===============

- Remove power notice signal
- Rely on GNOME session idle detection
- Proxy g-s inhibitors over to the session
- Read status message from session
- Accept a full path for theme (Matthias Clasen)

===============
Version 2.25.2
===============

- Fix pixmap leak (Matthias Clasen)
- Fix dialog border width

===============
Version 2.25.1
===============

- Add bevel around dialog
- Fixed #444927, get idle time from xorg
- Fixed #492949, screensavers-personal-slideshow should not show hidden directories (Jörg)
- Fixed #530561, "Random" preview is always the last saver (Maxim Ermilov)
- Fixed #561855, Calls wrong slideshow binary if there is one installed in $PATH (Frederic Crozat)

Translations:

src/gnome-screensaver-command.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*-
 *
 * Copyright (C) 2004-2006 William Jon McCann <mccann@jhu.edu>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 *
 * Authors: William Jon McCann <mccann@jhu.edu>
 *
 */

#include "config.h"

#include <stdlib.h>
#include <locale.h>

#include <glib.h>
#include <gio/gio.h>
#include <glib/gi18n.h>

#include "gs-bus.h"

static gboolean do_quit       = FALSE;
static gboolean do_lock       = FALSE;
static gboolean do_activate   = FALSE;
static gboolean do_deactivate = FALSE;
static gboolean do_version    = FALSE;

static gboolean do_query      = FALSE;
static gboolean do_time       = FALSE;

static GOptionEntry entries [] = {
        { "exit", 0, 0, G_OPTION_ARG_NONE, &do_quit,
          N_("Causes the screensaver to exit gracefully"), NULL },
        { "query", 'q', 0, G_OPTION_ARG_NONE, &do_query,
          N_("Query the state of the screensaver"), NULL },
        { "time", 't', 0, G_OPTION_ARG_NONE, &do_time,
          N_("Query the length of time the screensaver has been active"), NULL },
        { "lock", 'l', 0, G_OPTION_ARG_NONE, &do_lock,
          N_("Tells the running screensaver process to lock the screen immediately"), NULL },
        { "activate", 'a', 0, G_OPTION_ARG_NONE, &do_activate,
          N_("Turn the screensaver on (blank the screen)"), NULL },
        { "deactivate", 'd', 0, G_OPTION_ARG_NONE, &do_deactivate,
          N_("If the screensaver is active then deactivate it (un-blank the screen)"), NULL },
        { "version", 'V', 0, G_OPTION_ARG_NONE, &do_version,
          N_("Version of this application"), NULL },
        { NULL }
};

static GMainLoop *loop = NULL;

static GDBusMessage *
screensaver_send_message_bool (GDBusConnection *connection,
                               const char      *name,
                               gboolean         value)
{
        GDBusMessage *message, *reply;
        GError       *error;

        g_return_val_if_fail (connection != NULL, NULL);
        g_return_val_if_fail (name != NULL, NULL);

        message = g_dbus_message_new_method_call (GS_SERVICE,
                                                  GS_PATH,
                                                  GS_INTERFACE,
                                                  name);
        if (message == NULL) {
                g_warning ("Couldn't allocate the dbus message");
                return NULL;
        }

        g_dbus_message_set_body (message, g_variant_new ("(b)", value));

        error = NULL;
        reply = g_dbus_connection_send_message_with_reply_sync (connection,
                                                                message,
                                                                G_DBUS_SEND_MESSAGE_FLAGS_NONE,
                                                                -1,
                                                                NULL,
                                                                NULL,
                                                                &error);
        if (error != NULL) {
                g_warning ("unable to send message: %s", error->message);
                g_clear_error (&error);
        }

        g_dbus_connection_flush_sync (connection, NULL, &error);
        if (error != NULL) {
                g_warning ("unable to flush message queue: %s", error->message);
                g_clear_error (&error);
        }

        g_object_unref (message);

        return reply;
}

static GDBusMessage *
screensaver_send_message_void (GDBusConnection *connection,
                               const char      *name,
                               gboolean         expect_reply)
{
        GDBusMessage *message, *reply;
        GError       *error;

        g_return_val_if_fail (connection != NULL, NULL);
        g_return_val_if_fail (name != NULL, NULL);

        message = g_dbus_message_new_method_call (GS_SERVICE,
                                                  GS_PATH,
                                                  GS_INTERFACE,
                                                  name);
        if (message == NULL) {
                g_warning ("Couldn't allocate the dbus message");
                return NULL;
        }

        error = NULL;

        if (! expect_reply) {
                reply = NULL;

                g_dbus_connection_send_message (connection,
                                                message,
                                                G_DBUS_SEND_MESSAGE_FLAGS_NONE,
                                                NULL,
                                                &error);

        } else {
                reply = g_dbus_connection_send_message_with_reply_sync (connection,
                                                                        message,
                                                                        G_DBUS_SEND_MESSAGE_FLAGS_NONE,
                                                                        -1,
                                                                        NULL,
                                                                        NULL,
                                                                        &error);
        }

        if (error != NULL) {
                g_warning ("unable to send message: %s", error->message);
                g_clear_error (&error);
        }

        g_dbus_connection_flush_sync (connection, NULL, &error);
        if (error != NULL) {
                g_warning ("unable to flush message queue: %s", error->message);
                g_clear_error (&error);
        }

        g_object_unref (message);

        return reply;
}

static gboolean
screensaver_is_running (GDBusConnection *connection)
{
        GVariant *reply;
        gboolean exists = FALSE;

        g_return_val_if_fail (connection != NULL, FALSE);

        reply = g_dbus_connection_call_sync (connection,
                                             DBUS_SERVICE,
                                             DBUS_PATH,
                                             DBUS_INTERFACE,
                                             "GetNameOwner",
                                             g_variant_new ("(s)", GS_SERVICE),
                                             NULL,
                                             G_DBUS_CALL_FLAGS_NO_AUTO_START,
                                             -1,
                                             NULL,
                                             NULL);
        if (reply != NULL) {
                exists = TRUE;
                g_variant_unref (reply);
        }

        return exists;
}

static gboolean
do_command (GDBusConnection *connection)
{
        GDBusMessage *reply;

        if (do_quit) {
                reply = screensaver_send_message_void (connection, "Quit", FALSE);
                goto done;
        }

        if (do_query) {
                GVariant     *body;
                gboolean      v;

                if (! screensaver_is_running (connection)) {
                        g_message ("Screensaver is not running!");
                        goto done;
                }

                reply = screensaver_send_message_void (connection, "GetActive", TRUE);
                if (reply == NULL) {
                        g_message ("Did not receive a reply from the screensaver.");
                        goto done;
                }

                body = g_dbus_message_get_body (reply);
                g_variant_get (body, "(b)", &v);
                g_object_unref (reply);

                if (v)  {
                        g_print (_("The screensaver is active\n"));
                } else {
                        g_print (_("The screensaver is inactive\n"));
                }
        }

        if (do_time) {
                GVariant *body;
                gboolean  v;
                gint32    t;

                reply = screensaver_send_message_void (connection, "GetActive", TRUE);
                if (reply == NULL) {
                        g_message ("Did not receive a reply from the screensaver.");
                        goto done;
                }

                body = g_dbus_message_get_body (reply);
                g_variant_get (body, "(b)", &v);
                g_object_unref (reply);

                if (v) {
                        reply = screensaver_send_message_void (connection, "GetActiveTime", TRUE);
                        if (reply == NULL) {
                                g_message ("Did not receive a reply from the screensaver.");
                                goto done;
                        }

                        body = g_dbus_message_get_body (reply);
                        g_variant_get (body, "(u)", &t);
                        g_object_unref (reply);

                        g_print (ngettext ("The screensaver has been active for %d second.\n", "The screensaver has been active for %d seconds.\n", t), t);
                } else {
                        g_print (_("The screensaver is not currently active.\n"));
                }
        }

        if (do_lock) {
                reply = screensaver_send_message_void (connection, "Lock", FALSE);
                g_assert (reply == NULL);
        }

        if (do_activate) {
                reply = screensaver_send_message_bool (connection, "SetActive", TRUE);
                if (reply == NULL) {
                        g_message ("Did not receive a reply from the screensaver.");
                        goto done;
                }
                g_object_unref (reply);
        }

        if (do_deactivate) {
                reply = screensaver_send_message_bool (connection, "SetActive", FALSE);
                if (reply == NULL) {
                        g_message ("Did not receive a reply from the screensaver.");
                        goto done;
                }
                g_object_unref (reply);
        }

 done:
        g_main_loop_quit (loop);
        return FALSE;
}

int
main (int    argc,
      char **argv)
{
        GDBusConnection *connection;
        GOptionContext  *context;
        gboolean         retval;
        GError          *error;

#ifdef ENABLE_NLS
        bindtextdomain (GETTEXT_PACKAGE, GNOMELOCALEDIR);
# ifdef HAVE_BIND_TEXTDOMAIN_CODESET
        bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
# endif
        textdomain (GETTEXT_PACKAGE);
#endif

        g_type_init ();

        g_set_prgname (argv[0]);

        if (setlocale (LC_ALL, "") == NULL) {
                g_warning ("Locale not understood by C library, internationalization will not work\n");
        }

        error = NULL;
        context = g_option_context_new (NULL);
        g_option_context_add_main_entries (context, entries, NULL);
        retval = g_option_context_parse (context, &argc, &argv, &error);

        g_option_context_free (context);

        if (! retval) {
                g_warning ("%s", error->message);
                g_error_free (error);
                return EXIT_FAILURE;
        }

        if (do_version) {
                g_print ("%s %s\n", argv [0], VERSION);
                return EXIT_SUCCESS;
        }

        connection = g_bus_get_sync (G_BUS_TYPE_SESSION, NULL, &error);
        if (connection == NULL) {
                g_message ("Failed to get session bus: %s", error->message);
                g_error_free (error);
                return EXIT_FAILURE;
        }

        g_idle_add ((GSourceFunc) do_command, connection);

        loop = g_main_loop_new (NULL, FALSE);
        g_main_loop_run (loop);

        g_object_unref (connection);

        return EXIT_SUCCESS;
}

